AutoCollection/OutgoingHttpDependencies.ts

```diff
@@ -0,0 +1,128 @@
+                    https, options, ...requestArgs);
+                if (requestResult && options && !options[AutoCollectHttpDependencies.disableOutgoingHttpAutoCollectionOption]) {
+                    AutoCollectHttpDependencies.trackDependency(this._client, options, requestResult);
+                }
+                return requestResult;
+            };
+        }
+    }
+
+
+    // TODO: make fully specific
+    private static _isRequestOptions(requestOptions: string | http.RequestOptions | https.RequestOptions): boolean {
+        return ( requestOptions !== null && typeof requestOptions === 'string' || typeof requestOptions === 'object');
+    }
+    private static _isClientRequest(request: http.ClientRequest) : boolean {
+        return ( request !== null && typeof request === 'object');
+    }
+
+    /**
+     * Tracks an outgoing request. Because it may set headers this method must be called before
+     * writing content to or ending the request.
+     */
+    public static trackDependency(client: Client, requestOptions: string | http.RequestOptions | https.RequestOptions, request: http.ClientRequest,
+            properties?: { [key: string]: string }) {
+
+        if (!AutoCollectHttpDependencies._isRequestOptions(requestOptions) || !AutoCollectHttpDependencies._isClientRequest(request) || !client) {
+            Logging.info("AutoCollectHttpDependencies.trackDependency was called with invalid parameters: ", !requestOptions, !request, !client);
+            return;
+        }
+
+        // TODO(joshgav): get rootRequest (first param) from correlation context?
+        let outgoingHttpDependencyParser = new OutgoingHttpDependencyParser(null, client, requestOptions, request);
+
+        // Add the source ikey hash to the request headers, if a value was not already present in the request.
+        // The getHeader/setHeader methods aren't available on very old Node versions, and
+        // are not included in the v0.10 type declarations currently used. So check if the
+        // methods exist before invoking them.
+        // TODO(joshgav): Is this guaranteed to be called *before* sending the outgoing request?
+        if (request['getHeader'] && request['setHeader'] &&
+                !request['getHeader'](HttpHeaders.RemoteDependency.myIkey)) {
+            request['setHeader'](HttpHeaders.RemoteDependency.myIkey, outgoingHttpDependencyParser.myIkeyHash);
+        }
+
+        // Collect dependency telemetry about the request when it finishes.
+        if (request.on) {
+            request.on('response', (response: http.IncomingMessage) => {
+                outgoingHttpDependencyParser.onResponse(response, properties);
+                // TODO(joshgav): Is it okay to change ClientResponse -> IncomingMessage?
+                var context : { [name: string]: any; } = { "http.RequestOptions": requestOptions, "http.ClientRequest": request, "http.ClientResponse": response };
```

Comment:
Do you know when Node switched from ClientResponse to IncomingMessage? I see ClientResponse isn't documented anymore